src/frontend/src/expr/type_inference.rs

```diff
@@ -117,255 +117,191 @@ impl FuncSign {
     map: &mut HashMap<FuncSign, DataTypeName>,
     exprs: &[ExprType],
-    arg1: &[DataTypeName],
-    arg2: &[DataTypeName],
-    ret: DataTypeName,
+    args: &[DataTypeName],
 ) {
-    for (expr, a1, a2) in iproduct!(exprs, arg1, arg2) {
-        map.insert(FuncSign::new_binary(*expr, *a1, *a2), ret);
+    for e in exprs {
+        for (li, lt) in args.iter().enumerate() {
+            for (ri, rt) in args.iter().enumerate() {
+                let ret = if li <= ri { rt } else { lt };
+                map.insert(FuncSign::new(*e, vec![*lt, *rt]), *ret);
+            }
+        }
     }
 }
 
-fn build_commutative_binary_funcs(
+fn build_commutative_funcs(
     map: &mut HashMap<FuncSign, DataTypeName>,
-    exprs: &[ExprType],
-    arg1: &[DataTypeName],
-    arg2: &[DataTypeName],
+    expr: ExprType,
+    arg0: DataTypeName,
+    arg1: DataTypeName,
     ret: DataTypeName,
 ) {
-    build_binary_funcs(map, exprs, arg1, arg2, ret);
-    build_binary_funcs(map, exprs, arg2, arg1, ret);
-}
-
-fn build_ternary_funcs(
-    map: &mut HashMap<FuncSign, DataTypeName>,
-    exprs: &[ExprType],
-    arg1: &[DataTypeName],
-    arg2: &[DataTypeName],
-    arg3: &[DataTypeName],
-    ret: DataTypeName,
-) {
-    for (expr, a1, a2, a3) in iproduct!(exprs, arg1, arg2, arg3) {
-        map.insert(FuncSign::new_ternary(*expr, *a1, *a2, *a3), ret);
-    }
+    map.insert(FuncSign::new(expr, vec![arg0, arg1]), ret);
+    map.insert(FuncSign::new(expr, vec![arg1, arg0]), ret);
 }
 
 fn build_type_derive_map() -> HashMap<FuncSign, DataTypeName> {
     use {DataTypeName as T, ExprType as E};
     let mut map = HashMap::new();
-    let num_types = vec![
+    let all_types = [
+        T::Boolean,
         T::Int16,
         T::Int32,
         T::Int64,
+        T::Decimal,
         T::Float32,
         T::Float64,
-        T::Decimal,
+        T::Varchar,
+        T::Date,
+        T::Timestamp,
+        T::Timestampz,
+        T::Time,
+        T::Interval,
     ];
-    let all_types = vec![
+    let num_types = [
         T::Int16,
         T::Int32,
         T::Int64,
+        T::Decimal,
         T::Float32,
         T::Float64,
-        T::Boolean,
-        T::Varchar,
-        T::Decimal,
-        T::Time,
-        T::Timestamp,
-        T::Interval,
-        T::Date,
-        T::Timestampz,
     ];
-    let str_types = vec![T::Varchar];
-    let atm_exprs = vec![E::Add, E::Subtract, E::Multiply, E::Divide, E::Modulus];
-    let cmp_exprs = vec![
+
+    // logical expressions
+    for e in [E::Not, E::IsTrue, E::IsNotTrue, E::IsFalse, E::IsNotFalse] {
+        map.insert(FuncSign::new(e, vec![T::Boolean]), T::Boolean);
+    }
+    for e in [E::And, E::Or] {
+        map.insert(FuncSign::new(e, vec![T::Boolean, T::Boolean]), T::Boolean);
+    }
+
+    // comparison expressions
+    for e in [E::IsNull, E::IsNotNull] {
+        for t in all_types {
+            map.insert(FuncSign::new(e, vec![t]), T::Boolean);
+        }
+    }
+    let cmp_exprs = &[
```

Comment:
```suggestion
    for e in [E::IsNull, E::IsNotNull] {
        for t in all_types {
            map.insert(FuncSign::new(e, vec![t]), T::Boolean);
        }
    }
    // comparison expressions
    let cmp_exprs = &[
```

Reply:
These are also considered "comparison expressions", because they accept any types, compare against null, and return boolean. Similarly there are `Is[Not]DistinctFrom` but we do not support them yet.

In contrast, `Is[Not]{True,False}` only accept boolean as input.
https://www.risingwave.dev/docs/latest/sql-operators-functions/